source/common/router/config_impl.cc

```diff
@@ -193,6 +240,56 @@ const RouteEntry* RouteEntryImplBase::routeEntry() const {
   }
 }
 
+const Route* RouteEntryImplBase::clusterEntry(uint64_t random_value) const {
+  // Gets the route object chosen from the list of weighted clusters
+  // (if there is one) or returns self.
+  if (weighted_clusters_.empty()) {
+    return this;
+  }
+
+  uint64_t selected_value = random_value % WeightedClusterEntry::MAX_CLUSTER_WEIGHT;
+  uint64_t begin = 0UL;
+  uint64_t end = 0UL;
+
+  // Find the right cluster to route to based on the interval in which
+  // the selected value falls.  The intervals are determined as
+  // [0, cluster1_weight), [cluster1_weight, cluster1_weight+cluster2_weight),..
+  for (const WeightedClusterEntryPtr& cluster : weighted_clusters_) {
+    end = begin + cluster->clusterWeight();
+    if (((selected_value >= begin) && (selected_value < end)) ||
+        (end >= WeightedClusterEntry::MAX_CLUSTER_WEIGHT)) {
+      // end > WeightedClusterEntry::MAX_CLUSTER_WEIGHT : This case can only occur
+      // with Runtimes, when the user specifies invalid weights such that
+      // sum(weights) > WeightedClusterEntry::MAX_CLUSTER_WEIGHT.
+      // In this case, terminate the search and just return the cluster
+      // whose weight cased the overflow
+      return cluster.get();
+    }
+    begin = end;
+  }
+  return nullptr;
```

Comment:
I think I already commented on this, but this should never happen, switch to NOT_IMPLEMENTED